pkgs/applications/misc/dmenu-rs/default.nix

```diff
@@ -0,0 +1,111 @@
+{ lib
+, fetchFromGitHub
+, rustPlatform
+, clang
+, libclang
+, llvm
+, m4
+, pkg-config
+, expat
+, fontconfig
+, libX11
+, libXft
+, libXinerama
+, python3
+}:
+
+# The dmenu-rs package has extensive plugin support. However, this derivation
+# only builds the package with the default set of plugins. If you'd like to
+# further customize dmenu-rs you can build it from the source.
+# See: https://github.com/Shizcow/dmenu-rs#plugins
+rustPlatform.buildRustPackage rec {
+  pname = "dmenu-rs";
+  version = "5.5.1";
+
+  src = fetchFromGitHub {
+    owner = "Shizcow";
+    repo = pname;
+    rev = version;
+    sha256 = "sha256-WpDqBjIZ5ESnoRtWZmvm+gNTLKqxL4IibRVCj0yRIFM=";
+  };
+
+  nativeBuildInputs = [
+    clang
+    libclang
+    llvm
+    m4
+    pkg-config
+    python3
+  ];
+
+  buildInputs = [
+    expat
+    fontconfig
+    libX11
+    libXft
+    libXinerama
+  ];
+
+  # The dmenu-rs repository does not include a Cargo.lock because of its
+  # dynamic build and plugin support. Generating it with make and checking it
+  # in to nixpkgs here was the easiest way to supply it to buildRustPackage.
+  # See: https://github.com/Shizcow/dmenu-rs/issues/34#issuecomment-757415584
+  cargoLock = {
+    lockFile = ./Cargo.lock;
+  };
+
+  # Several nixpkgs' functions from its rust build tooling expect cargo files
+  # to exist at the project's root directory unless otherwise specified.
+  cargoRoot = "src";
+
+  # The rust-xcb dependency dynamically generates rust code at build time.
+  # This derivation uses nixpkgs rust functions that vendor each cargo
+  # dependency's source code into the READ-ONLY nix store. To avoid the code
+  # generation step failing, we copy the rust-xcb source out of the nix store
+  # and make it writeable.
+  # See: https://github.com/rust-x-bindings/rust-xcb/tree/v0.8.2
+  patchPhase = ''
+    chmod +w /build/cargo-vendor-dir
+    mkdir -p /build/cargo-vendor-dir/xcb-0.8.2-readwrite
+    cp --recursive /build/cargo-vendor-dir/xcb-0.8.2/. /build/cargo-vendor-dir/xcb-0.8.2-readwrite
+    unlink /build/cargo-vendor-dir/xcb-0.8.2
+    mv /build/cargo-vendor-dir/xcb-0.8.2-readwrite /build/cargo-vendor-dir/xcb-0.8.2
+    chmod -R +w /build/cargo-vendor-dir/xcb-0.8.2/
+  '';
+
+  # As suggested in the nixpkgs manual section on building rust packages, the
+  # derivation's Cargo.lock file is copied into the source.
+  postPatch = ''
+    cp ${./Cargo.lock} src/Cargo.lock
+  '';
+
+  preConfigure = ''
+    sed -i "s@PREFIX = /usr/local@PREFIX = $out@g" config.mk
+  '';
+
+  buildPhase = ''
+    make
+  '';
+
+  installPhase = ''
+    make install
+  '';
```

Comment:
I thought these looked fishy at first, but it actually does make sense. We want the rust platform stuff because it does the vendoring and all that, but it doesn't just do `make` and `make install` like `stdenv.mkDerivation` does by default.

That said, this is very much *not* going to cross compile correctly. Not critical, but unfortunate, and probably fixable.

Reply:
I've added a comment that tries to explain these phases' surprising appearance.

I'll be honest, I don't know the exact consequences of this package not cross compiling correctly, but it sure sounds bad. Is it related to not running `./configure` in the `configurePhase`? Should I go read up on Autoconf and figure out how to add that into this derivation?

Reply:
I wouldn't worry about it not cross compiling; frankly I don't think you can cross compile a desktop environment with nixpkgs right now anyway. It was just something I found noteworthy. But basically the package's makefile is hostile to cross compilation because e.g. it hardcodes `CC=cc` and doesn't have a way to choose the cargo target. It would need (relatively trivial) changes to their build scripting to fix this.

Reply:
Thanks for calljng it out. I'm not really a contributor to `dmenu-rs` (I fixed 1 typo) but it's something I'll keep in mind and who knows maybe I can fix it upstream some time in the future.

Reply:
did you consider using mkDerivation with cargoSetupHook to handle fetching sources?

to quote the manual:
> The `rustPlatform.buildRustPackage` function is split into several hooks: cargoSetupHook to set up vendoring for Cargo-based projects, cargoBuildHook to build a project using Cargo, cargoInstallHook to install a project using Cargo, and cargoCheckHook to run tests in Cargo-based projects. With this change, mixed-language projects can use the relevant hooks within builders other than `buildRustPackage`. However, these changes also required several API changes to `buildRustPackage` itself:

Reply:
I didn't. That sounds like a more natural way to write this derivation -- as a standard make-based build that happens to include cargo deps. I'll take a stab at rewriting it that way later today. Thanks for this!